### PR TITLE
[WIP]SQL extension: expire snapshot

### PR DESCRIPTION
**Pull request by liukun4515**
support the expire snapshot and some sql test
syntax:

```
CALL catalog.schema.expire_snapshots(
  namespace  => 'namespace_name' 
  table => 'table_name',
  older_than => timestamp,  
  retain_last => number
)

The namespace/table/older_than is required.
The retain_last is optional.
```

issue: [#1597](https://github.com/apache/iceberg/issues/1597)

### CODE REVIEW

**Comment by aokolnychyi** [comment]
@liukun4515, could you rebase this one? There were a few changes since the draft PR.

**Comment by aokolnychyi** [comment]
Also, per design doc, `older_than` is optional and should be defaulted to `System.currentTimeMillis()` if not set.

Then we can keep only a given number of snapshots:
```
CALL catalog.schema.expire_snapshots(
  namespace => 'namespace_name',
  table => 'table_name',
  retain_last => 100
)
```

File: spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveProcedures.scala [hunk 1]
```diff
@@ -50,7 +50,8 @@ case class ResolveProcedures(spark: SparkSession) extends Rule[LogicalPlan] with
 
     // optional params should be at the end
     procedure.parameters.sliding(2).foreach {
-      case Array(previousParam, currentParam) if previousParam.required && !currentParam.required =>
+          // Optional arg must be behind the required arg.
+      case Array(previousParam, currentParam) if !previousParam.required && currentParam.required =>
```
Comment (rdblue) [inline spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveProcedures.scala:54]: I think this has already been fixed.

File: spark3-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestExpireSnapshotProcedure.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        ImmutableList.of(row(retainedLastNum, currentTimeStamp)),
+        output);
+    Assert.assertEquals(3, Lists.newArrayList(table.snapshots()).size());
+    Assert.assertEquals(tailSnapshot.snapshotId(), Lists.newArrayList(table.snapshots()).get(0).snapshotId());
+    Assert.assertEquals(headSnapshot.snapshotId(), Lists.newArrayList(table.snapshots()).get(2).snapshotId());
+  }
+
+  @Test
+  public void testExpireSnapshotUsingNameArgs() {
+
```
Comment (rdblue) [inline spark3-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestExpireSnapshotProcedure.java:100]: Will need these tests, I think.

File: spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveProcedures.scala (hunk 1 above)
Comment (liukun4515) [reply spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveProcedures.scala:54]: i will rebase the master code.

File: spark3-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestExpireSnapshotProcedure.java (hunk 2 above)
Comment (liukun4515) [reply spark3-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestExpireSnapshotProcedure.java:100]: yes, i will add the test case

**Comment by liukun4515** [comment]
> Also, per design doc, `older_than` is optional and should be defaulted to `System.currentTimeMillis()` if not set.
> 
> Then we can keep only a given number of snapshots:
> 
> ```
> CALL catalog.schema.expire_snapshots(
>   namespace => 'namespace_name',
>   table => 'table_name',
>   retain_last => 100
> )
> ```

The the default `older_than`(`System.currentTime`) is a very dangerous.

If some one not give the `older_than` value, all the snapshots will be dropped.

@chenjunjiedada

**Comment by rdblue** [comment]
> If don't give the older_than value, all the snapshots will be dropped.

The current snapshot is never dropped, but you're right that it is easy to drop lots of data doing this.

We've thought about introducing table properties to default these. Seems like a good time to add them. I'm thinking some time interval for keeping snapshots (so expireOlderThan would default to `System.currentTimeMillis() - interval`) and some default number of snapshots to keep.

@aokolnychyi, do you agree?

**Comment by aokolnychyi** [comment]
The design doc assumed that while both `retain_last` and `older_than` are optional, at least one of them is required to prevent the situation described above. That being said, we can probably support `NOW()` literals later to just keep a given number of snapshots. So maybe we don't have to make `older_than` optional.

@rdblue, I am +1 for adding table props to control the default behavior. I was saving this until we have an idea how a generic VACUUM command is going to look like but we can probably introduce properties for retention right now.

**Comment by aokolnychyi** [comment]
I think we can add table props in a follow-up PR.

**Comment by rdblue** [comment]
If we want to add table config in a follow-up, then I think we should go ahead and use a reasonable default here. If `older_than` is not set, then let's just default it to `System.currentTimeMillis() - (3 days in millis)`. I think that's a good default and we can fix the default interval with a table property after this is done.

**Comment by RussellSpitzer** [comment]
Something that this brings up for me, is how does a user discover the
information about these defaults? Particularly from a full SQL context, I
wonder if it makes sense to create a "Help" command which prints out this
info or returns it as a dataframe. Something like

```help expireSnapshots```

**Comment by rdblue** [comment]
> Something that this brings up for me, is how does a user discover the information about these defaults?

I think that we should probably add `docs()` to the stored procedure interface and add a command to show them, like `HELP system.expire_snapshots`. Probably not `DESCRIBE` since that is used for tables and views.

**Comment by aokolnychyi** [comment]
If we are going to add defaults, I'd say let's go ahead and define table properties now.

**Comment by rdblue** [comment]
@aokolnychyi, what about `history.expire.snapshot-age-ms` and `history.expire.min-snapshots`? Those would control the age after which a snapshot can expire (the interval for `expireOlderThan(now - interval)`) and the retain-last as minimum number of snapshots to keep.

**Comment by aokolnychyi** [comment]
@rdblue @liukun4515, I've submitted PR #1868 for table properties. Please, take a look.

**Comment by aokolnychyi** [comment]
I took commits done by @liukun4515 from this PR and submitted #1874 with up-to-date code.

**Comment by liukun4515** [comment]
> I took commits done by @liukun4515 from this PR and submitted #1874 with up-to-date code.

thanks, i will close the pull request.

**Pull request closed without merging**